Spigot-Server-Patches/0594-Significantly-improve-performance-of-the-end-generat.patch

```diff
@@ -0,0 +1,58 @@
+From: SuperCoder7979 <k.pranav@gmail.com>
+Date: Tue, 3 Nov 2020 23:48:05 -0600
+Subject: [PATCH] Significantly improve performance of the end generation
+
+This patch implements a noise cache for the end which significantly reduces the computation time of generation. This results in about a 3x improvement.
+
+Original code by SuperCoder7979 and Gegy in Lithium, licensed under LGPL-3.0 <Source: https://github.com/jellysquid3/lithium-fabric>
+
+Co-authored-by: Gegy <gegy1000@gmail.com>
+Co-authored-by: Dylan Xaldin <Puremin0rez515@gmail.com>
+Co-authored-by: pop4959 <pop4959@gmail.com>
+
+diff --git a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+index 9c15cd303ccc2928be4e3e7080e74e8314d7e0f2..eb370c226d401ad0d1b29f054b34e6cfb5d192c1 100644
+--- a/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
++++ b/src/main/java/net/minecraft/server/WorldChunkManagerTheEnd.java
+@@ -3,6 +3,7 @@ package net.minecraft.server;
+ import com.google.common.collect.ImmutableList;
+ import com.mojang.serialization.Codec;
+ import com.mojang.serialization.codecs.RecordCodecBuilder;
++import it.unimi.dsi.fastutil.HashCommon; // Paper
+ import java.util.List;
+ 
+ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+@@ -22,6 +23,8 @@ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+     private final BiomeBase k;
+     private final BiomeBase l;
+     private final BiomeBase m;
++    private static final long[] keys = new long[8192]; // Paper
++    private static final float[] values = new float[8192]; // Paper
+ 
+     public WorldChunkManagerTheEnd(IRegistry<BiomeBase> iregistry, long i) {
+         this(iregistry, i, (BiomeBase) iregistry.d(Biomes.THE_END), (BiomeBase) iregistry.d(Biomes.END_HIGHLANDS), (BiomeBase) iregistry.d(Biomes.END_MIDLANDS), (BiomeBase) iregistry.d(Biomes.SMALL_END_ISLANDS), (BiomeBase) iregistry.d(Biomes.END_BARRENS));
+@@ -81,8 +84,21 @@ public class WorldChunkManagerTheEnd extends WorldChunkManager {
+                 long i2 = (long) (k + k1);
+                 long j2 = (long) (l + l1);
+ 
+-                if (i2 * i2 + j2 * j2 > 4096L && noisegenerator3handler.a((double) i2, (double) j2) < -0.8999999761581421D) {
+-                    float f1 = (MathHelper.e((float) i2) * 3439.0F + MathHelper.e((float) j2) * 147.0F) % 13.0F + 9.0F;
++                // Paper start - Significantly improve end generation performance by using a noise cache
++                long key = ChunkCoordIntPair.pair((int) i2, (int) j2);
++                int index = (int) HashCommon.mix(key) & 8191;
++                float f1 = Long.MIN_VALUE;
++                if (keys[index] == key) {
++                    f1 = values[index];
++                } else {
++                    if (i2 * i2 + j2 * j2 > 4096L && noisegenerator3handler.a((double) i2, (double) j2) < -0.8999999761581421D) {
```

Comment:
how often is it <= 4096? If this happens often, we should skip cache logic for this 

like
 if (<= 4096)  f1 = minvalue
 else if (incache) 

so that the cache isn't getting clobbered by static values

Reply:
It should only be <= 4096 when generating the main end island and the surrounding ~1000 blocks of void 😅